### ISSUE TITLE
`indent` rule: multiple consecutive comments lead to increased indentation requirement

### ISSUE THREAD

**Issue by rasenplanscher** [eslint/eslint#11340]
**Tell us about your environment**

* **ESLint Version: 5.12.0**
* **Node Version: 11.6.0**
* **npm Version: 6.5.0-next.0**

**What parser (default, Babel-ESLint, etc.) are you using?**
default

**Please show your full configuration:**

<details>
<summary>Configuration</summary>

`test.eslintrc.js`
```js
module.exports = {
	rules: {
		indent: [
			'error',
			'tab',
		],
	},
}
```

</details>

**What did you do? Please include the actual source code causing the issue, as well as the command that you used to run ESLint.**

In order to auto-generate documentation for a project of mine, I added JSdoc comments.
I'm exporting a class and in order to not have it listed as "index" by `documentation`, I wedge the comments in between the export and the class declaration.
The constructor is over-loaded, so I added multiple comments for it.
All in all, that leads to the following structure:

`test.js`
```js
module.exports =
/**
 * variant 1
 */
/**
 * variant 2
 */
/**
 * variant 3
 */
/**
 * variant 4
 */
/**
 * variant 5
 */
class Test {}
```

```bash
npx eslint test.js -c .\test.eslintrc.js
```

**What did you expect to happen?**
No error, exit code 0

**What actually happened? Please include the actual, raw output from ESLint.**
```
test.js
   5:1  error  Expected indentation of 1 tab but found 0  indent
   8:1  error  Expected indentation of 1 tab but found 0  indent
  11:1  error  Expected indentation of 1 tab but found 0  indent

✖ 3 problems (3 errors, 0 warnings)
  3 errors and 0 warnings potentially fixable with the `--fix` option.
```

The extra indentation requirement starts happening with three comments, and it affects all but the first and last comments.

Also, the first and last comments need to have at least three lines each in order to "count".
That means, removing the middle line of the first comment will remove the for lint error.

**Are you willing to submit a pull request to fix this bug?**
Yes, but I would appreciate pointers to where to start digging.

**Comment by rasenplanscher** [comment]
May or may not be related to #10884

**Comment by platinumazure** [comment]
Hi @rasenplanscher, thanks for the issue.

The current comment indentation logic is as follows:

A comment can be indented in up to 3 valid locations:

1. If there is a line of code above it, with no blank lines, the comment can align with that line of code.
1. If there is a line of code below it, with no blank lines, the comment can align with that line of code.
1. Otherwise, it must be aligned in the "correct" location (i.e., where a line of code would be indented).

The first comment matches the first case. The last comment matches the second case. The other comments would have the third case applied, and the indent rule by default expects hanging code underneath an assignment to be indented by one level.

----------

Do you see an error for the indentation of the class expression? I would have expected to see a lint error there (i.e., it should also be indented by 1 tab by default), but I don't see an error from that line in your report.

----------

I think you've probably hit a corner case with the current comment indent logic. It might be that we want to allow comments to be aligned relative to other comments immediately above or below them as well. I'm not 100% sure.

I'm also a big confused about why the comments have to be 3 lines tall. It seems to me that this should occur with block comments of any length/height, as well as line comments. I might have to dig into that at some point.

**Comment by rasenplanscher** [comment]
> 
> 
> Hi @rasenplanscher, thanks for the issue.
> 
> The current comment indentation logic is as follows:
> 
> A comment can be indented in up to 3 valid locations:
> 
>     1. If there is a line of code above it, with no blank lines, the comment can align with that line of code.
> 
>     2. If there is a line of code below it, with no blank lines, the comment can align with that line of code.
> 
>     3. Otherwise, it must be aligned in the "correct" location (i.e., where a line of code would be indented).

That certainly explains the difference between first/last and middle comments 🤔

> 
> 
> The first comment matches the first case. The last comment matches the second case. The other comments would have the third case applied, and the indent rule by default expects hanging code underneath an assignment to be indented by one level.
> 
> Do you see an error for the indentation of the class expression? I would have expected to see a lint error there (i.e., it should also be indented by 1 tab by default), but I don't see an error from that line in your report.

I cannot confirm that. The following runs without an issue:
```
module.exports =
class Test {}
```

I have posted the complete configuration and the complete error output.

Additionally, I don't see anything on https://eslint.org/docs/rules/indent indicating that assignments are checked the way you describe. If they were, I would expect a configuration option for that, because this is likely to have many supporters for either strategy.

> 
> I think you've probably hit a corner case with the current comment indent logic. It might be that we want to allow comments to be aligned relative to other comments immediately above or below them as well. I'm not 100% sure.

I think, comments should generally be treated the same as everything else, as far as feasible. While I (and presumably many others) do align them with each other, that's more or less incidental because I want them at their proper levels. Consequently, I see no reason for a change in that direction, beyond the current issue.

> 
> I'm also a big confused about why the comments have to be 3 lines tall. It seems to me that this should occur with block comments of any length/height, as well as line comments. I might have to dig into that at some point.

On that we are in full agreement 👍
